components/sensor/ina219/ina219.go

```diff
@@ -0,0 +1,229 @@
+	}
+	return deps, nil
+}
+
+func init() {
+	registry.RegisterComponent(
+		sensor.Subtype,
+		modelname,
+		registry.Component{Constructor: func(
+			ctx context.Context,
+			deps registry.Dependencies,
+			config config.Component,
+			logger golog.Logger,
+		) (interface{}, error) {
+			attr, ok := config.ConvertedAttributes.(*AttrConfig)
+			if !ok {
+				return nil, rdkutils.NewUnexpectedTypeError(AttrConfig{}, config.ConvertedAttributes)
+			}
+			return newSensor(ctx, deps, config.Name, attr, logger)
+		}})
+
+	config.RegisterComponentAttributeMapConverter(sensor.Subtype, modelname,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf AttrConfig
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		}, &AttrConfig{})
+}
+
+func newSensor(
+	ctx context.Context,
+	deps registry.Dependencies,
+	name string,
+	attr *AttrConfig,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+	b, err := board.FromDependencies(deps, attr.Board)
+	if err != nil {
+		return nil, fmt.Errorf("ina219 init: failed to find board: %w", err)
+	}
+	localB, ok := b.(board.LocalBoard)
+	if !ok {
+		return nil, fmt.Errorf("board %s is not local", attr.Board)
+	}
+	i2cbus, ok := localB.I2CByName(attr.I2CBus)
+	if !ok {
+		return nil, fmt.Errorf("ina219 init: failed to find i2c bus %s", attr.I2CBus)
+	}
+	addr := attr.I2cAddr
+	if addr == 0 {
+		addr = defaultI2Caddr
+		logger.Infof("using i2c address : %d", defaultI2Caddr)
+	}
+
+	s := &ina219{
+		logger: logger,
+		bus:    i2cbus,
+		addr:   byte(addr),
+	}
+
+	err = s.calibrate(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// ina219 is a i2c sensor device that reports voltage, current and power
+type ina219 struct {
+	generic.Unimplemented
+	logger     golog.Logger
+	bus        board.I2C
+	addr       byte
+	currentLSB int64
+	powerLSB   int64
+	cal        uint16
+}
+
+type PowerMonitor struct {
+	Shunt   int64
+	Voltage float64
+	Current float64
+	Power   float64
+}
+
+func (d *ina219) calibrate(ctx context.Context) error {
+	if senseResistor <= 0 {
+		return fmt.Errorf("ina219 calibrate: senseResistor value invalid %d", senseResistor)
+	}
+	if maxCurrent <= 0 {
+		return fmt.Errorf("ina219 calibrate: maxCurrent value invalid %d", maxCurrent)
+	}
+
+	d.currentLSB = maxCurrent / (1 << 15)
+	d.powerLSB = int64((maxCurrent*20 + (1 << 14)) / (1 << 15))
+	// Calibration Register = 0.04096 / (current LSB * Shunt Resistance)
+	// Where lsb is in Amps and resistance is in ohms.
+	// Calibration register is 16 bits.
+	cal := calibratescale / (d.currentLSB * int64(senseResistor))
+	if cal >= (1 << 16) {
+		return fmt.Errorf("ina219 calibrate: calibration register value invalid %d", cal)
+	}
+	d.cal = uint16(cal)
+
+	return nil
+}
+
+// Readings returns a list containing three items (voltage, current, and power).
+func (d *ina219) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
```

Comment:
Curious: do you know the latency of this call? fine if not, just curious.

Reply:
>  latency of this call?

All I2C interactions use the board's clock, and the sensor has no way to delay its response. I don't remember the default I2C clock speed, but it's on the order of 100 kilohertz. Just doing a back-of-the-envelope calculation, we're sending/receiving roughly 25 bytes (roughly 200 bits) over a 100 kilobaud channel, so once you're able to open the handle on the bus, I'd expect the rest to take about 2 milliseconds. (The time needed for opening the handle will depend on how many other things want to also use the bus; it's basically the time needed to lock a mutex).